flamingpy/noise/cv.py

```diff
@@ -0,0 +1,649 @@
+        star_p_val = macro_graph.nodes[vertex]["hom_val_p"]
+
+        # Types of state for the four micronodes directly neighboring
+        # macronode j.
+        types = [
+            macro_graph.nodes[neighbor]["state"] if neighbor else None for neighbor in neighbors
+        ]
+        # Phase error probability and number of p-squeezed states
+        # among the four micronodes in the vicinity of macronode j
+        p_err = 0
+        num_p = 0
+        outcome = 2 * star_p_val
+        gkp_inds = []
+        for i in (1, 2, 3, 4):
+            if types[i - 1] == "p":
+                num_p += 1
+                outcome -= Z_arr[i]
+            if types[i - 1] == "GKP":
+                if delta > 0:
+                    p_err += Z_err_cond(2 * delta, Z_arr[i], use_hom_val=True)
+                gkp_inds += [i]
+        if delta > 0:
+            p_err += Z_err_cond(2 * (2 + num_p) * delta, outcome, use_hom_val=True)
+        p_err = min(p_err, 0.5)
+        p_err = max(p_err, 0)
+
+        bitp = GKP_binner([outcome])[0]
+        bitq = GKP_binner(Z_arr[gkp_inds].astype(np.float64)) if gkp_inds else 0
+
+        processed_bit_val = (bitp + np.sum(bitq)) % 2
+
+        # Update the reduced RHG lattice with the effective
+        # homodyne value and the phase error probability.
+        central_vert = tuple(round(i) for i in vertex)
+        self.reduced_graph.nodes[central_vert]["bit_val"] = processed_bit_val
+        self.reduced_graph.nodes[central_vert]["p_phase_cond"] = p_err
+
+    def reduce(self, noise_model):
+        """Reduce the macronode lattice macro_graph to the canonical
+        reduced_graph.
+
+        Follow the procedure in arXiv:2104.03241. Take the macronode lattice
+        macro_graph, apply noise based on noise_layer and noise_model, designate
+        micronodes as planets and stars, conduct homodyne measurements, process
+        these measurements, and compute conditional phase error probabilities.
+
+        Modify reduced_graph into a canonical lattice with effective measurement
+        outcomes and phase error probabilities stored as node attributes.
+
+        Args:
+            macro_graph (EGraph): the macronode lattice
+            reduced_graph (EGraph): the reduced lattice
+            noise_layer (CVLayer): the noise layer
+            noise_model (dict): the dictionary of noise parameters to be fed into
+                noise_layer
+            bs_network (np.array): the beamsplitter network used to entangle the
+                macronodes.
+
+        Returns:
+            None
+        """
+        macro_graph = self.egraph
+        # Sample for the initial state
+        self._apply_initial_noise(noise_model)
+        self._permute_indices_and_label()
+        self._entangle_states()
+        self._measure_syndrome()
+        # Process homodyne outcomes and calculate phase error probabilities
+        for j in range(0, len(macro_graph) - 3, 4):
+            self._reduce_jth_macronode(j)
```

Comment:
Looking at this, it doesn't seem hard to me to have a `macronize` function taking a `CVLayer` to a `CVMacroLayer` if we want to.

Reply:
Could you explain what you mean? (Note there have been several changes since your comment)